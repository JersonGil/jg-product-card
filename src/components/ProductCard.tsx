import React, { createContext, CSSProperties } from 'react'
import styles from '../styles/styles.module.css'
import { useProducts } from '../hooks/useProduct'
import {
  ProductContextProps,
  Product,
  onChangeArgs,
  InitialValues,
  ProductCardHandlers,
} from '../interfaces/interfaces'

export interface Props {
  product: Product;
  className?: string;
  style?: CSSProperties;
  value?: number;
  initialValues?: InitialValues;
  onChange?: (args: onChangeArgs) => void;
  children: (args: ProductCardHandlers) => JSX.Element;
}

export const ProductContext = createContext({} as ProductContextProps)

const { Provider } = ProductContext

export const ProductCard = ({
  product,
  className,
  style,
  value,
  initialValues,
  onChange,
  children,
}: Props) => {
  const { counter, increaseBy, maxCounter, isMaxCountReach, reset } =
    useProducts({
      onChange,
      product,
      value,
      initialValues,
    })

  return (
    <Provider value={{ product, counter, increaseBy, maxCounter }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReach,
          product,
          maxCount: initialValues?.maxCount,
          reset,
          increaseBy
        })}
      </div>
    </Provider>
  )
}
