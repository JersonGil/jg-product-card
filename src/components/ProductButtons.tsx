import React, { useContext, CSSProperties, useCallback } from 'react'
import styles from '../styles/styles.module.css'
import { ProductContext } from './ProductCard'

export interface Props {
  className?: string;
  style?: CSSProperties;
}

export const ProductButtons = ({ className }: Props) => {
  const { increaseBy, counter, maxCounter } = useContext(ProductContext)

  const isMaxReached = useCallback(() => {
    if (maxCounter) return counter >= maxCounter
    return
  }, [counter, maxCounter])

  return (
    <div className={`${styles.buttonsContainer} ${className}`}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        {' '}
        -{' '}
      </button>
      <div className={styles.countLabel}> {counter} </div>
      <button
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
        disabled={!!isMaxReached()}
        onClick={() => increaseBy(1)}
      >
        {' '}
        +{' '}
      </button>
    </div>
  )
}
