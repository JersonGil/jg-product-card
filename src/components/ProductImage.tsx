import React, { useContext, CSSProperties } from 'react'
import noImage from '../assets/no-image.jpg'
import styles from '../styles/styles.module.css'
import { ProductContext } from './ProductCard'

export interface Props {
  img?: string;
  className?: string;
  style?: CSSProperties;
}

export const ProductImage = ({ img, className }: Props) => {
  const { product } = useContext(ProductContext)
  let imgToshow: string

  if (img) {
    imgToshow = img
  } else if (product.img) {
    imgToshow = product.img
  } else {
    imgToshow = noImage
  }

  return (
    <img
      className={`${styles.productImg} ${className}`}
      src={imgToshow}
      alt='product'
    />
  )
}
