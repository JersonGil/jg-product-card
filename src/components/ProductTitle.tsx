import React, { useContext, CSSProperties } from 'react'
import styles from '../styles/styles.module.css'
import { ProductContext } from './ProductCard'

export interface Props {
  className?: string;
  title?: string;
  style?: CSSProperties;
}

export const ProductTitle = ({ title, className }: Props) => {
  const { product } = useContext(ProductContext)

  return (
    <span className={`${styles.productDescription} ${className}`}>
      {title ? title : product.title}
    </span>
  )
}
