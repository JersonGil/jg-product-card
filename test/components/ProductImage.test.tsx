import React from 'react'
import renderer from 'react-test-renderer'
import { ProductImage, ProductCard } from '../../src/components'
import { product1 } from '../data/products'

describe('ProductImage', () => {
  test('debe mostrar el componente con su imagen', () => {
    const wrapper = renderer.create(
      <ProductImage img={product1.img} />
    )

    expect(wrapper.toJSON()).toMatchSnapshot()
  })

  test('debe mostrar el componente con su imagen del producto 1', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {
          () => (
            <ProductImage/>
          )
        }
      </ProductCard>
    )

    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
