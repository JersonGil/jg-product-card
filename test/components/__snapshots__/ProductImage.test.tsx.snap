// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage debe mostrar el componente con su imagen 1`] = `
<img
  alt="product"
  className="productImg undefined"
  src="./coffee-mug.png"
/>
`;

exports[`ProductImage debe mostrar el componente con su imagen del producto 1 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
